<a href="https://www.kaggle.com/code/alinahurtova/practice-1-nn?scriptVersionId=353878276" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

Імпорт модулів

In [1]:
import torch

# Завдання 1

Перегляньте інформацію про доступні процесори, та виведіть про них інформацію

In [2]:
print(torch.cuda.is_available())  # True/False

if torch.cuda.is_available():
    print(torch.cuda.device_count())  # Кількість GPU
    print(torch.cuda.get_device_name(0))  # Назва першого GPU

tensor = torch.tensor([1, 2, 3])
print(tensor.device)  # CPU

True
2
Tesla T4
cpu


# Завдання 2
Підключіться до процесора GPU, та знову виведіть інформацію про доступні процесори

In [3]:
tensor_gpu = tensor.to("cuda")

tensor_gpu

tensor([1, 2, 3], device='cuda:0')

# Завдання 3
Створіть випадковий тензор розміру (2, 3). Виведіть інформацію про його:

* розмір
* тип даних
* процесор
* наявність градієнта

In [4]:
tensor = torch.randn(2,3)

In [5]:
tensor

tensor([[-0.2000,  0.2090, -0.1972],
        [-0.1561,  0.6230, -1.2521]])

In [6]:
tensor.dtype

torch.float32

In [7]:
tensor.shape

torch.Size([2, 3])

In [8]:
tensor.device

device(type='cpu')

# Завдання 4

Створіть 2 тензора розміру 100. Виведіть
* суму векторів
* частку векторів(перед цим замініть один елемент на 0)
* максимальне та мінімальне значення(методи .min() .max())
* середнє значення(метод .mean())
* квадрад різниці
* середнє значення квадрату різниці

In [9]:
tensor1=torch.randn(100)
tensor2=torch.randn(100)

In [10]:
tensor1

tensor([ 1.0725, -2.0613,  0.9316, -0.5120, -0.5221, -0.6049,  0.3167,  1.0592,
        -0.1890,  1.6656, -0.0732, -1.6323, -0.3820,  1.7129,  1.2932,  2.2495,
         0.5569,  0.0244, -0.5701,  0.2228,  1.0406,  0.1382, -1.0992,  0.4270,
         0.5200,  0.5273, -0.2723, -1.4134,  1.0247, -0.9057,  0.3047, -1.1059,
         0.5573,  0.2089, -1.3150, -0.3353,  0.7066, -0.9122,  1.0402, -0.1753,
         0.1366,  0.0923,  1.6729,  0.3339, -1.9506, -0.6739, -1.0730, -0.5710,
         0.2806,  0.2392,  0.9253,  0.2947,  1.1053,  0.1184,  0.3966,  0.4313,
         1.0622,  1.2779,  0.8819,  1.8560, -0.9547,  2.4395,  1.3583,  0.0493,
         1.7928, -0.9608, -0.6320,  0.2522, -0.5679,  0.0250,  2.4255,  0.4979,
        -0.9941,  1.8259, -0.1566, -0.1314,  1.5575,  0.4929, -0.3862,  1.4454,
         0.0141, -0.2402,  0.8777,  1.1376,  0.9362, -1.6723,  0.4371, -0.7996,
        -1.2963,  1.3711,  0.4999,  1.6322,  0.7905, -0.5919, -0.9999,  0.7414,
        -0.5510,  0.1692, -0.7285,  0.92

In [11]:
tensor2

tensor([ 0.4400, -1.3129, -0.0499,  0.9216,  1.6155, -0.0440, -0.0578,  1.6037,
        -0.0471, -0.7963,  1.5796,  0.4929,  1.8440, -0.7144, -0.4329, -0.3518,
        -1.5318,  0.5279, -0.0050,  0.2226,  0.9432,  0.1677,  0.1732,  0.6516,
         0.0670,  1.7505,  0.1964,  0.2076, -0.9879,  0.7321, -0.1899, -1.2168,
         0.3761, -0.9476,  0.9206,  0.2103,  2.2883,  1.3300,  0.3799, -0.0385,
        -0.1527,  1.1786,  0.7320,  1.5599,  2.0853, -0.8304,  3.3785,  0.0383,
         0.1146, -0.4705, -0.2980,  1.9898,  0.3446, -1.4017,  0.9353,  0.0502,
         0.1454,  1.1857,  0.0102,  0.5384, -0.2552,  0.4522, -0.2121,  0.6424,
         1.4833, -0.8775,  0.8319,  1.7052,  0.5086,  1.0165,  0.2959, -1.9350,
        -1.5617, -0.8290, -0.6589,  0.3658, -1.9243, -0.0488, -0.1579, -1.5247,
        -0.5491, -1.7732,  1.2578, -0.3754,  2.3325,  1.7948, -0.4480, -0.7745,
         0.6592,  1.3678, -0.4248,  1.0959,  0.7180,  0.2630,  0.4069, -1.2163,
         0.6849, -0.8143,  1.0121,  0.46

In [12]:
tensor2[0] = 0
tensor_div = tensor1/tensor2

In [13]:
tensor_min = tensor.min()
tensor_max = tensor.max()

In [14]:
tensor_sum = tensor1+ tensor2


In [15]:
tensor_max

tensor(0.6230)

In [16]:
tensor_min

tensor(-1.2521)

In [17]:
tensor_div

tensor([        inf,  1.5700e+00, -1.8674e+01, -5.5550e-01, -3.2322e-01,
         1.3758e+01, -5.4808e+00,  6.6048e-01,  4.0093e+00, -2.0916e+00,
        -4.6347e-02, -3.3115e+00, -2.0715e-01, -2.3976e+00, -2.9874e+00,
        -6.3945e+00, -3.6357e-01,  4.6283e-02,  1.1462e+02,  1.0008e+00,
         1.1032e+00,  8.2400e-01, -6.3476e+00,  6.5535e-01,  7.7598e+00,
         3.0125e-01, -1.3863e+00, -6.8091e+00, -1.0373e+00, -1.2372e+00,
        -1.6040e+00,  9.0886e-01,  1.4816e+00, -2.2046e-01, -1.4284e+00,
        -1.5945e+00,  3.0880e-01, -6.8583e-01,  2.7383e+00,  4.5511e+00,
        -8.9501e-01,  7.8306e-02,  2.2852e+00,  2.1407e-01, -9.3538e-01,
         8.1161e-01, -3.1761e-01, -1.4898e+01,  2.4490e+00, -5.0844e-01,
        -3.1050e+00,  1.4812e-01,  3.2074e+00, -8.4476e-02,  4.2402e-01,
         8.5925e+00,  7.3043e+00,  1.0778e+00,  8.6179e+01,  3.4474e+00,
         3.7405e+00,  5.3951e+00, -6.4043e+00,  7.6693e-02,  1.2087e+00,
         1.0950e+00, -7.5970e-01,  1.4788e-01, -1.1

In [18]:
tensor_diff = tensor1-tensor2

In [19]:
tensor_diff_sq = tensor_diff**2


tensor_diff_sq

tensor([1.1503e+00, 5.6011e-01, 9.6327e-01, 2.0552e+00, 4.5694e+00, 3.1461e-01,
        1.4028e-01, 2.9646e-01, 2.0116e-02, 6.0611e+00, 2.7317e+00, 4.5166e+00,
        4.9549e+00, 5.8918e+00, 2.9794e+00, 6.7666e+00, 4.3626e+00, 2.5346e-01,
        3.1934e-01, 3.4900e-08, 9.4833e-03, 8.7136e-04, 1.6190e+00, 5.0436e-02,
        2.0522e-01, 1.4962e+00, 2.1974e-01, 2.6277e+00, 4.0507e+00, 2.6823e+00,
        2.4465e-01, 1.2298e-02, 3.2817e-02, 1.3376e+00, 4.9980e+00, 2.9758e-01,
        2.5016e+00, 5.0275e+00, 4.3606e-01, 1.8716e-02, 8.3697e-02, 1.1800e+00,
        8.8521e-01, 1.5029e+00, 1.6288e+01, 2.4472e-02, 1.9816e+01, 3.7132e-01,
        2.7567e-02, 5.0373e-01, 1.4966e+00, 2.8731e+00, 5.7863e-01, 2.3106e+00,
        2.9021e-01, 1.4526e-01, 8.4054e-01, 8.5083e-03, 7.5974e-01, 1.7361e+00,
        4.8929e-01, 3.9495e+00, 2.4660e+00, 3.5182e-01, 9.5808e-02, 6.9462e-03,
        2.1430e+00, 2.1113e+00, 1.1588e+00, 9.8307e-01, 4.5350e+00, 5.9193e+00,
        3.2216e-01, 7.0483e+00, 2.5223e-

In [20]:
tensor_diff_sq.mean()

tensor(2.1897)

In [21]:
tensor_div.mean()

tensor(inf)

# Завдання 5

У вашому розпорядженні є нейронна мережа, яка прогнозує **орієнтовну ціну квартири** на основі двох **ознак**:

* **площі квартири** (у квадратних метрах);
* **кількості кімнат**.

Вам надано готовий клас `NeuralNetwork` із методом `predict`.

1. Створіть екземпляр класу `NeuralNetwork`, передайте як параметр `device`.
2. За допомогою методу `predict()` визначте прогнозовану ціну квартири, яка:
   * має площу **65 м²**;
   * має **2 кімнати**.
     
3. Для цього створіть тензор із відповідними числами та скористайтесь методом `predict()`.
4. Усі нейромережі очікують дані розміру `(N, K)`, де:
   * `N` — кількість об'єктів, для яких виконується прогноз;
   * `K` — кількість ознак. У цьому випадку їх дві — **площа квартири та кількість кімнат**.

5. Спробуйте скористатись методом `predict()` разом із `torch.no_grad()`:

```python
with torch.no_grad():
    price = model.predict(...)
```

**Завдання:** отримайте прогнозовану ціну квартири та виведіть результат на екран.


In [22]:
import torch


class NeuralNetwork:
    def __init__(self, device="cpu"):
        self.device = device

        self.W = torch.tensor(
            [[3.3, -38.]],
            requires_grad=True,
            device=device
        )

        self.b = torch.tensor(
            [176.],
            requires_grad=True,
            device=device
        )

    def predict(self, X):
        if not isinstance(X, torch.Tensor):
            raise TypeError("X повинен бути об'єктом torch.Tensor.")

        if X.ndim != 2:
            raise ValueError(
                f"Очікується двовимірний тензор, отримано {X.shape}."
            )

        if X.shape[1] != 2:
            raise ValueError(
                f"Очікується 2 ознаки, отримано {X.shape[1]}."
            )

        if not X.is_floating_point():
            raise TypeError(
                "Тензор повинен містити числа з плаваючою комою (float32 або float64)."
            )

        if X.device != self.W.device:
            raise RuntimeError(
                f"X знаходиться на {X.device}, а модель на {self.W.device}."
            )

        return X @ self.W.T + self.b


In [23]:
# Ваш код
neural_network = NeuralNetwork(device = 'cuda')


In [24]:
tensor = torch.tensor([65, 2])

In [25]:
tensor

tensor([65,  2])

In [26]:
tensor1 = tensor.reshape(1,2)

In [27]:
tensor1.shape

torch.Size([1, 2])

In [28]:
tensor.shape

torch.Size([2])

In [29]:
tensor1 = tensor1.float()

In [30]:
tensor1 = tensor1.to('cuda')

tensor1

tensor([[65.,  2.]], device='cuda:0')

In [31]:
neural_network.predict(tensor1)

tensor([[314.5000]], device='cuda:0', grad_fn=<AddBackward0>)

In [32]:
tensor1.device

device(type='cuda', index=0)

# **Завдання 6**

У таблиці наведено інформацію про три квартири.

| **Квартира** | **Площа (м²)** | **Кількість кімнат** |
| ------------ | -------------: | -------------------: |
| Квартира 1   |             40 |                    1 |
| Квартира 2   |             60 |                    3 |
| Квартира 3   |             100 |                    2 |

Виконайте наступні кроки:

1. Створіть **три окремі тензори** (`X1`, `X2`, `X3`), кожен з яких містить дані про одну квартиру.
2. За допомогою методу `predict()` обчисліть **прогнозовану ціну** для кожної квартири окремо.
3. Створіть **один тензор** `X` розміру **(3, 2)**, де:
   * кожен рядок відповідає одній квартирі;
   * перший стовпець — площа квартири;
   * другий стовпець — кількість кімнат.

4. За допомогою одного виклику `predict(X)` отримайте **прогнозовану ціну одразу для всіх трьох квартир**.
5. Виведіть розмір тензору з трьома прогнозованими цінами, який ви отримали на кроці 4.
6. Порівняйте результати, отримані окремими викликами `predict()` та одним спільним викликом `predict(X)`.


In [33]:
# Створіть три окремі тензори (X1, X2, X3), кожен з яких містить дані про одну квартиру.

x1 = torch.tensor([40.0, 1.0])
x2 = torch.tensor([60.0, 3.0])
x3 = torch.tensor([100.0, 2.0])

In [34]:
x1 = x1.to('cuda')
x2 = x2.to('cuda')
x3 = x3.to('cuda')

x1

tensor([40.,  1.], device='cuda:0')

In [35]:
# 1 - одна квартира, 2 - кількість ознак
x1 = x1.reshape(1,2)
x2 = x2.reshape(1,2)
x3 = x3.reshape(1,2)

x1.shape

torch.Size([1, 2])

In [36]:
# За допомогою методу predict() обчисліть прогнозовану ціну для кожної квартири окремо.

neural_network.predict(x1)

tensor([[270.]], device='cuda:0', grad_fn=<AddBackward0>)

In [37]:
neural_network.predict(x2)

tensor([[260.]], device='cuda:0', grad_fn=<AddBackward0>)

In [38]:
neural_network.predict(x3)

tensor([[430.]], device='cuda:0', grad_fn=<AddBackward0>)

In [39]:

X = torch.tensor([
    [40.0, 1.0], 
    [60.0, 3.0],
    [100.0, 2.0]
])

In [40]:
X = X.to('cuda')


In [41]:
neural_network.predict(X)

tensor([[270.],
        [260.],
        [430.]], device='cuda:0', grad_fn=<AddBackward0>)

# Завдання 7

Три квартири були продані, і стали відомі їхні **реальні ціни**.

| Квартира   | Реальна ціна (1000$) |
| ---------- | ---------------: |
| Квартира 1 |           210 |
| Квартира 2 |          310 |
| Квартира 3 |          360 |

Нехай прогноз моделі збережений у змінній `y_pred`, а реальні значення — у `y_true`.

Визначте, **наскільки модель помилилася**, використавши два різних підходи.

### Варіант 1

1. Знайдіть різницю між прогнозованою та реальною ціною (`y_pred - y_true`).
2. Обчисліть середнє значення цієї різниці за допомогою `torch.mean()`.

### Варіант 2

1. Знайдіть різницю між прогнозованою та реальною ціною (`y_pred - y_true`).
2. Позбудьтеся від'ємних значень за допомогою `torch.abs()`.
3. Обчисліть середнє значення отриманих помилок.

### Запитання

1. Який із двох варіантів краще характеризує якість моделі? Поясніть свою відповідь.
2. Напишіть функцію `loss_fn()`, яка реалізує обраний варіант.

```python
def loss_fn(y_pred, y_true):
    """
    Обчислює середню абсолютну помилку (Mean Absolute Error, MAE).

    Параметри:
        y_pred (torch.Tensor): прогнозовані значення моделі.
        y_true (torch.Tensor): реальні (правильні) значення.

    Повертає:
        torch.Tensor: середню абсолютну помилку.
    """
    ...
```




In [42]:
def loss_fn(y_pred, y_true):
    """
    Обчислює середню абсолютну помилку (Mean Absolute Error, MAE).

    Параметри:
        y_pred (torch.Tensor): прогнозовані значення моделі.
        y_true (torch.Tensor): реальні (правильні) значення.

    Повертає:
        torch.Tensor: середню абсолютну помилку.
    """
    MAE = y_pred - y_true
    return MAE.abs().mean()

In [43]:
y_true = torch.tensor([210.0, 310.0, 360.0])


y_true = y_true.to('cuda')

y_true

tensor([210., 310., 360.], device='cuda:0')

In [44]:
y_true = y_true.reshape(3,1)
y_true

tensor([[210.],
        [310.],
        [360.]], device='cuda:0')

In [45]:
y_pred = neural_network.predict(X)

y_pred

tensor([[270.],
        [260.],
        [430.]], device='cuda:0', grad_fn=<AddBackward0>)

In [46]:
MAE = loss_fn(y_pred, y_true)

MAE

tensor(60., device='cuda:0', grad_fn=<MeanBackward0>)

# Завдання 8

У вашому розпорядженні є вчитель, який може навчити нейронну мережу робити точніші прогнози **цін на квартири**.

Клас `Teacher` вже реалізований. Під час створення об'єкта необхідно передати модель, яку потрібно навчати:

```python
teacher = Teacher(model)
```

### Завдання

1. Створіть об'єкт класу `Teacher`, передавши йому модель.
2. Виведіть тензор `X`, який містить дані про квартири:
   * площу квартири;
   * кількість кімнат.

3. Виведіть тензор `y_true` з **реальними цінами квартир**.
4. Отримайте прогноз моделі `y_pred`.
5. Виведіть прогнозовані ціни квартир.
6. За допомогою власної функції `loss_fn()` обчисліть помилку моделі.
7. Виведіть значення помилки.
8. Скористайтеся методом `teach()` класу `Teacher`, передавши йому обчислену помилку.
9. Після навчання знову отримайте прогноз `y_pred`.
10. Виведіть нові прогнозовані ціни квартир.
11. Знову обчисліть та виведіть помилку.
12. Зробіть висновок: **чи стала модель прогнозувати ціни квартир точніше після одного кроку навчання?**



In [47]:
import torch


class Teacher:
    def __init__(self, model, lr=0.02):
        """
        Створює вчителя для навчання моделі.

        Параметри:
            model (NeuralNetwork): модель, яку потрібно навчати.
            lr (float): швидкість навчання.
        """
        self.model = model

        self.optimizer = torch.optim.SGD(
            [model.W, model.b],
            lr=lr
        )

    def teach(self, loss):
        """
        Виконує один крок навчання.

        Параметри:
            loss (torch.Tensor): значення функції втрат.
        """
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()


Алгоритм
X → прогноз → помилка → перенавчання моделі → backward → зміна W,b → новий прогноз → нова помилка

In [48]:
# 1. Створюємо Teacher, передаємо модель
teacher = Teacher(neural_network)

In [49]:
# 2. Виводимо дані про квартири
print("X (flats):")
print(X)

X (flats):
tensor([[ 40.,   1.],
        [ 60.,   3.],
        [100.,   2.]], device='cuda:0')


In [50]:
# 3. Реальні ціни
print("y_true:")
print(y_true)

y_true:
tensor([[210.],
        [310.],
        [360.]], device='cuda:0')


In [51]:
# 4. Прогноз моделі до навчання
y_pred = neural_network.predict(X)

In [52]:
print("Прогноз до навчання:")
print(y_pred)

Прогноз до навчання:
tensor([[270.],
        [260.],
        [430.]], device='cuda:0', grad_fn=<AddBackward0>)


In [53]:
# 5. Обчислюємо помилку до навчання
loss = loss_fn(y_pred, y_true)

print("Початкова помилка:")
print(loss)

Початкова помилка:
tensor(60., device='cuda:0', grad_fn=<MeanBackward0>)


In [54]:
# 6. Навчаємо модель
teacher.teach(loss)

In [55]:
# 7. Прогноз після навчання
y_pred = neural_network.predict(X)

print("Прогноз після навчання:")
print(y_pred)

Прогноз після навчання:
tensor([[248.6600],
        [227.9933],
        [376.6600]], device='cuda:0', grad_fn=<AddBackward0>)


In [56]:
# 8.Помилка після навчання
new_loss = loss_fn(y_pred, y_true)

print("Помилка після навчання:")
print(new_loss)

Помилка після навчання:
tensor(45.7756, device='cuda:0', grad_fn=<MeanBackward0>)


Модель стала краще прогнозувати після першого кроку навчання - помилка зменшилася

Початкова помилка:
tensor(60., device='cuda:0', grad_fn=<MeanBackward0>)

Помилка після навчання:
tensor(45.7756, device='cuda:0', grad_fn=<MeanBackward0>)


# Завдання 9

Одного кроку навчання недостатньо, щоб модель навчилася добре прогнозувати ціни. Тому повторіть процес навчання **100 разів**.

### Завдання

1. Створіть об'єкт класу `Teacher`, передавши йому модель.
2. Створіть тензор `X`, який містить дані про квартири.
3. Створіть тензор `y_true`, який містить реальні реальні ціни.
4. Створіть цикл, який виконається **100 разів**.
5. На кожній ітерації циклу:
   * отримайте прогноз моделі `y_pred`;
   * обчисліть помилку за допомогою функції `loss_fn()`;
   * виведіть помилку на екран;
   * передайте помилку в метод `teacher.teach()`, щоб виконати один крок навчання.
    
6. Після завершення циклу ще раз отримайте прогноз моделі.
7. Виведіть:
   * прогнозовані заробітні плати;
   * значення функції втрат.
    
8. Порівняйте отримані прогнози з реальними ціни. Зробіть висновок, чи покращилася точність моделі після 100 кроків навчання.
9. Cпробуйте зменшити навчальний крок в `teacher`

In [57]:
# Створюємо Teacher
teacher = Teacher(neural_network)

In [58]:
# Дані про квартири
X = torch.tensor([
    [40.0, 1.0],
    [60.0, 3.0],
    [100.0, 2.0]
]).to('cuda')

In [59]:
# Реальні ціни
y_true = torch.tensor([
    [210.0],
    [310.0],
    [360.0]
]).to('cuda')

In [60]:
# Навчання 100 разів 

losses = []

for i in range(100):

    y_pred = neural_network.predict(X)

    loss = loss_fn(y_pred, y_true)
    
    print(loss)

    losses.append(loss.item())

    teacher.teach(loss)

print("Початкова помилка:", losses[0])
print("Фінальна помилка:", losses[-1])

tensor(45.7756, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(56.0045, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8867, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.3023, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(59.9978, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7733, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(55.9356, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8844, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.2334, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(59.9955, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7711, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(55.8667, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8822, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.1645, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(59.9933, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7688, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(55.7979, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8799

In [61]:
# Робимо фінальний прогноз на останній навченій версії
y_pred = neural_network.predict(X)

y_pred

tensor([[249.5928],
        [231.0593],
        [378.6587]], device='cuda:0', grad_fn=<AddBackward0>)

In [62]:
# Фінальна помилку
final_loss = loss_fn(y_pred, y_true)

print("\nПрогнозовані ціни:")
print(y_pred)

print("\nРеальні ціни:")
print(y_true)

print("\nФінальна помилка:")
print(final_loss)


Прогнозовані ціни:
tensor([[249.5928],
        [231.0593],
        [378.6587]], device='cuda:0', grad_fn=<AddBackward0>)

Реальні ціни:
tensor([[210.],
        [310.],
        [360.]], device='cuda:0')

Фінальна помилка:
tensor(45.7307, device='cuda:0', grad_fn=<MeanBackward0>)


In [64]:
# Cпробуйте зменшити навчальний крок в teacher
# Створюємо нову нейронну мережу 
new_neural_network = NeuralNetwork(device="cuda")

In [65]:
# teacher з меншим навчальним кроком
teacher = Teacher(new_neural_network, lr=0.005)

In [66]:
# Дані про квартири
X = torch.tensor([
    [40.0, 1.0],
    [60.0, 3.0],
    [100.0, 2.0]
]).to('cuda')

In [67]:
# Реальні ціни
y_true = torch.tensor([
    [210.0],
    [310.0],
    [360.0]
]).to('cuda')

In [68]:
# Навчання 100 разів 

losses = []

for i in range(100):

    y_pred = new_neural_network.predict(X)

    loss = loss_fn(y_pred, y_true)
    
    print(loss)

    losses.append(loss.item())

    teacher.teach(loss)

print("Початкова помилка:", losses[0])

tensor(60., device='cuda:0', grad_fn=<MeanBackward0>)
tensor(56.4439, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(52.8878, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(49.3317, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7756, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(42.2194, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.3367, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(43.9972, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(42.6611, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7750, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(42.2189, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.3195, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(43.9967, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(42.6439, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.7744, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(42.2183, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(45.3023, device='cuda:0', grad_fn=<MeanBackward0>)
tensor(43.9961, de

In [69]:
# Робимо фінальний прогноз на останній навченій версії
y_pred = new_neural_network.predict(X)

y_pred

tensor([[243.5466],
        [220.7199],
        [363.7997]], device='cuda:0', grad_fn=<AddBackward0>)

In [70]:
# Фінальна помилку
final_loss = loss_fn(y_pred, y_true)

print("\nПрогнозовані ціни:")
print(y_pred)

print("\nРеальні ціни:")
print(y_true)

print("\nФінальна помилка:")
print(final_loss)


Прогнозовані ціни:
tensor([[243.5466],
        [220.7199],
        [363.7997]], device='cuda:0', grad_fn=<AddBackward0>)

Реальні ціни:
tensor([[210.],
        [310.],
        [360.]], device='cuda:0')

Фінальна помилка:
tensor(42.2088, device='cuda:0', grad_fn=<MeanBackward0>)


Фінальна помилака зменшилася 

з Фінальна помилка:
tensor(45.7307, device='cuda:0', grad_fn=<MeanBackward0>)

до Фінальна помилка:
tensor(42.2088, device='cuda:0', grad_fn=<MeanBackward0>)

# Завдання 10

У вашому розпорядженні є нейронна мережа, яка прогнозує **ймовірність дощу**.

Модель використовує три ознаки:

* температура повітря (°C);
* вологість повітря (%);
* швидкість вітру (км/год).

Результатом роботи моделі є **два числа**:

1. ймовірність того, що **дощу не буде**;
2. ймовірність того, що **дощ буде**.

Скористайтеся методом `predict()`, щоб визначити ймовірності для таких погодних умов:

* температура — **20°C**;
* вологість — **80%**;
* швидкість вітру — **15 км/год**.



In [ ]:
import torch


class RainNeuralNetwork:
    def __init__(self, device="cpu"):
        self.device = device

        self.W = torch.tensor(
            [
                [0.03, 0.01, 0.02],
                [0.06, -0.01, 0.08]
            ],
            requires_grad=True,
            device=device
        )

        self.b = torch.tensor(
            [-0.2, -0.3],
            requires_grad=True,
            device=device
        )

    def predict(self, X):
        if not isinstance(X, torch.Tensor):
            raise TypeError("X повинен бути об'єктом torch.Tensor.")

        if X.ndim != 2:
            raise ValueError(
                f"Очікується двовимірний тензор, отримано {X.shape}."
            )

        if X.shape[1] != 3:
            raise ValueError(
                f"Очікується 3 ознаки, отримано {X.shape[1]}."
            )

        if not X.is_floating_point():
            raise TypeError(
                "Тензор повинен містити числа з плаваючою комою."
            )

        if X.device != self.W.device:
            raise RuntimeError(
                f"X знаходиться на {X.device}, а модель на {self.W.device}."
            )

        logits = X @ self.W.T + self.b
        return torch.softmax(logits, dim=1)

# Завдання 11

Метеостанція надала прогноз погоди на **4 дні**.

| День | Температура (°C) | Вологість (%) | Швидкість вітру (км/год) |
| ---- | ---------------: | ------------: | -----------------------: |
| 1    |               30 |            40 |                        5 |
| 2    |               22 |            85 |                       18 |
| 3    |               15 |            95 |                       25 |
| 4    |               28 |            60 |                       10 |

Виконайте наступні кроки:

1. Створіть один тензор `X` розміру **(4, 3)**, де кожен рядок містить дані про один день.

2. За допомогою методу `predict()` отримайте ймовірності для кожного дня `y_pred`.

3. Скористайтеся функцією `torch.argmax(..., dim=1)`, щоб визначити прогноз моделі `result`:

   * `0` — дощу не буде;
   * `1` — дощ буде.

4. Виведіть:

   * тензор із ймовірностями `y_pred`;
   * тензор із прогнозами `result`


# Завдання 12

Метеорологи повідомили, чи справді пішов дощ у кожен із чотирьох днів.

| День | Дощ |
| ---- | :-: |
| 1    |  Ні |
| 2    | Так |
| 3    | Так |
| 4    |  Ні |

Для навчання моделі використайте готову функцію втрат **CrossEntropyLoss** з бібліотеки **PyTorch**.

### Завдання

1. Створіть об'єкт функції втрат:

   ```python
   loss_fn = torch.nn.CrossEntropyLoss()
   ```

2. Створіть об'єкт класу `Teacher`, передавши йому модель.

3. Створіть тензор `X`, який містить погодні умови за всі 4 дні.

4. Створіть тензор `y_true`, який містить правильні відповіді:

   * `0` — дощу не було;
   * `1` — дощ був.

5. Створіть цикл, який виконається **100 разів**.

6. На кожній ітерації циклу:

   * отримайте ймовірності від моделі `y_pred`:
   * обчисліть помилку:
   * виведіть помилку
   * передайте помилку методу `teacher.teach(loss)` для виконання одного кроку навчання.

7. Після завершення циклу ще раз отримайте прогноз моделі.

8. За допомогою функції `torch.argmax(..., dim=1)` отримайте прогноз моделі:

   * `0` — дощу не буде;
   * `1` — дощ буде.

9. Виведіть:

   * ймовірності, отримані від моделі;
   * прогнози моделі (`0` або `1`);
   * правильні відповіді
   * значення функції втра


# Завдання 13

Попробуйте натренувати модель використовуючи функцію втрат, яку ви раніше застосовували для задачі визначення цін на квартири (`torch.nn.MSELoss()`), замість `CrossEntropyLoss`.

Метеорологи повідомили, чи справді пішов дощ у кожен із чотирьох днів.

| День | Дощ |
|------|-----|
| 1    | Ні  |
| 2    | Так |
| 3    | Так |
| 4    | Ні  |

**Завдання**

1. Створіть об'єкт класу `Teacher`, передавши йому модель.
2. Створіть тензор `X`, який містить погодні умови за всі 4 дні.
3. Створіть тензор `y_true` розміром (4, 2), де кожен рядок містить інформацію про те чи йшов дощ в цей день:
   - `[1, 0]` — дощу не було
   - `[0, 1]` — дощ був
   -  формат інший тому що ваша функція вимагає щоб `y_true` та `y_pred` були однакового розміру
4. Створіть цикл, який виконається 100 разів.

На кожній ітерації циклу:
- отримайте ймовірності від моделі `y_pred`;
- обчисліть помилку;
- виведіть помилку;
- передайте помилку методу `teacher.teach(loss)` для виконання одного кроку навчання.

6. Після завершення циклу ще раз отримайте прогноз моделі.
7. За допомогою функції `torch.argmax(..., dim=1)` отримайте прогноз моделі:
   - 0 — дощу не буде;
   - 1 — дощ буде.

**Виведіть:**
- ймовірності, отримані від моделі;
- прогнози моделі (0 або 1);
- правильні відповіді;
- значення функції втрат.
